File: ios/LayoutManager.swift
import CoreGraphics
import Foundation

final class LayoutManager {
  private let core = NitroListLayoutCore()

  struct EngagedRange {
    let start: Int
    let end: Int
    let version: Int
  }


  func setItemCount(_ count: Int) -> Bool {
    core.setItemCount(Int32(count))
  }

  func setEstimate(_ value: CGFloat) -> Bool {
    core.setEstimate(Float(value))
  }

  func setMeasurementEpsilon(_ value: CGFloat) {
    core.setMeasurementEpsilon(Float(value))
  }

  func setDirectionalBuffers(_ enabled: Bool) {
    core.setDirectionalBuffersEnabled(enabled)
  }

  func resetScrollVelocity() {
    core.resetScrollVelocity()
  }

  func setEstimatesFrozen(_ frozen: Bool) -> Bool {
    core.setEstimatesFrozen(frozen)
  }

  func setTypeAverages(_ enabled: Bool) {
    core.setTypeAveragesEnabled(enabled)
  }

  func setItemTypes(_ types: UnsafePointer<UInt16>?, count: Int) {
    core.setItemTypes(types, count: Int32(count))
  }

  func fillLayoutSlab(
    _ out: UnsafeMutablePointer<Double>,
    capacity: Int,
    scrollOffset: CGFloat,
    viewportHeight: CGFloat,
    drawDistance: CGFloat
  ) -> Int {
    Int(
      core.fillLayoutSlab(
        out,
        capacity: Int32(capacity),
        scrollOffset: Float(scrollOffset),
        viewportHeight: Float(viewportHeight),
        drawDistance: Float(drawDistance),
        outputScale: 1
      ))
  }

  func setItemSize(_ index: Int, _ size: CGFloat) -> Bool {
    core.setItemSize(Int32(index), size: Float(size))
  }

  func setItemSizes(_ pairs: UnsafePointer<Double>, count: Int) -> Bool {
    core.setItemSizesBatch(pairs, pairCount: Int32(count))
  }

  func seedTypeMeans(_ pairs: UnsafePointer<Double>, count: Int) -> Bool {
    core.seedTypeMeans(pairs, pairCount: Int32(count))
  }

  func setItemSizesAnchored(_ pairs: UnsafePointer<Double>, count: Int, anchorIndex: Int) -> CGFloat {
    CGFloat(
      core.setItemSizesBatchAnchored(pairs, pairCount: Int32(count), anchorIndex: Int32(anchorIndex)))
  }

  func resetItemSizes() -> Bool {
    core.resetItemSizes()
  }

  func remapItemSizes(_ pairs: UnsafePointer<Double>, count: Int) -> Bool {
    core.remapItemSizes(pairs, pairCount: Int32(count))
  }

  func resetAll() {
    core.resetAll()
  }

  func memoryFootprint() -> Int {
    Int(core.memoryFootprint())
  }


  func getTotalSize() -> CGFloat {
    CGFloat(core.totalSize())
  }

  func getOffset(_ index: Int) -> CGFloat {
    CGFloat(core.offset(for: Int32(index)))
  }

  func getSize(_ index: Int) -> CGFloat {
    CGFloat(core.size(for: Int32(index)))
  }

  func getLayoutVersion() -> Int {
    Int(core.layoutVersion())
  }

  func getEngagedRange(scrollOffset: CGFloat, viewportHeight: CGFloat, drawDistance: CGFloat) -> EngagedRange {
    let range = core.engagedRange(
      withScrollOffset: Float(scrollOffset),
      viewportHeight: Float(viewportHeight),
      drawDistance: Float(drawDistance)
    )
    return EngagedRange(start: Int(range.start), end: Int(range.end), version: Int(range.version))
  }
}

File: android/src/main/java/com/margelo/nitro/nitrolist/views/LayoutManager.kt
package com.margelo.nitro.nitrolist.views

import com.facebook.proguard.annotations.DoNotStrip
import java.nio.DoubleBuffer
import java.nio.ShortBuffer

@DoNotStrip
class LayoutManager {
  private val nativeHandle: Long = nativeCreate()

  private val rangeScratch = IntArray(3)

  fun setItemCount(count: Int): Boolean = nativeSetItemCount(nativeHandle, count)

  fun setEstimate(estimate: Float): Boolean = nativeSetEstimate(nativeHandle, estimate)

  fun setMeasurementEpsilon(epsilonPx: Float) = nativeSetMeasurementEpsilon(nativeHandle, epsilonPx)

  fun setDirectionalBuffers(enabled: Boolean) = nativeSetDirectionalBuffers(nativeHandle, enabled)

  fun resetScrollVelocity() = nativeResetScrollVelocity(nativeHandle)

  fun setEstimatesFrozen(frozen: Boolean): Boolean = nativeSetEstimatesFrozen(nativeHandle, frozen)

  fun setTypeAverages(enabled: Boolean) = nativeSetTypeAverages(nativeHandle, enabled)

  fun setItemTypes(types: ShortBuffer?, count: Int) = nativeSetItemTypes(nativeHandle, types, count)

  fun setColumnCount(columns: Int) = nativeSetColumnCount(nativeHandle, columns)

  fun setItemSpans(spans: ShortBuffer?, count: Int): Boolean =
    nativeSetItemSpans(nativeHandle, spans, count)

  fun fillLayoutSlab(
    slab: DoubleBuffer,
    capacityDoubles: Int,
    scrollOffsetPx: Float,
    viewportHPx: Float,
    drawDistancePx: Float,
    outputScale: Float,
  ): Int = nativeFillLayoutSlab(
    nativeHandle, slab, capacityDoubles, scrollOffsetPx, viewportHPx, drawDistancePx, outputScale,
  )

  fun fillTypeStats(out: DoubleBuffer, capacityDoubles: Int, outputScale: Float): Int =
    nativeFillTypeStats(nativeHandle, out, capacityDoubles, outputScale)

  fun setItemSize(index: Int, sizePx: Float): Boolean = nativeSetItemSize(nativeHandle, index, sizePx)

  fun setItemSizesDp(pairs: DoubleBuffer, count: Int, density: Float): Boolean =
    nativeSetItemSizesBatch(nativeHandle, pairs, count, density)

  fun setItemSizesDpAnchored(pairs: DoubleBuffer, count: Int, density: Float, anchorIndex: Int): Float =
    nativeSetItemSizesBatchAnchored(nativeHandle, pairs, count, density, anchorIndex)

  fun seedTypeMeansDp(pairs: DoubleBuffer, count: Int, density: Float): Boolean =
    nativeSeedTypeMeans(nativeHandle, pairs, count, density)

  fun resetItemSizes(): Boolean = nativeResetItemSizes(nativeHandle)

  fun remapItemSizes(pairs: DoubleBuffer, count: Int): Boolean =
    nativeRemapItemSizes(nativeHandle, pairs, count)

  fun resetAll() = nativeResetAll(nativeHandle)

  fun memoryFootprint(): Long = nativeGetMemoryFootprint(nativeHandle)

  fun getTotalSizePx(): Float = nativeGetTotalSize(nativeHandle)

  fun getOffsetPx(index: Int): Float = nativeGetOffset(nativeHandle, index)

  fun getSizePx(index: Int): Float = nativeGetSize(nativeHandle, index)

  fun getLayoutVersion(): Int = nativeGetLayoutVersion(nativeHandle)

  data class EngagedRange(val start: Int, val end: Int, val version: Int)

  fun getEngagedRange(scrollOffsetPx: Float, viewportHPx: Float, drawDistancePx: Float): EngagedRange =
    synchronized(rangeScratch) {
      nativeGetEngagedRange(nativeHandle, scrollOffsetPx, viewportHPx, drawDistancePx, rangeScratch)
      EngagedRange(rangeScratch[0], rangeScratch[1], rangeScratch[2])
    }

  @Suppress("deprecation")
  protected fun finalize() {
    nativeDestroy(nativeHandle)
  }

  private external fun nativeCreate(): Long
  private external fun nativeDestroy(handle: Long)
  private external fun nativeSetItemCount(handle: Long, count: Int): Boolean
  private external fun nativeSetEstimate(handle: Long, value: Float): Boolean
  private external fun nativeSetMeasurementEpsilon(handle: Long, value: Float)
  private external fun nativeSetDirectionalBuffers(handle: Long, enabled: Boolean)
  private external fun nativeResetScrollVelocity(handle: Long)
  private external fun nativeSetEstimatesFrozen(handle: Long, frozen: Boolean): Boolean
  private external fun nativeSetTypeAverages(handle: Long, enabled: Boolean)
  private external fun nativeSetItemTypes(handle: Long, types: ShortBuffer?, count: Int)
  private external fun nativeSetColumnCount(handle: Long, columns: Int)
  private external fun nativeSetItemSpans(handle: Long, spans: ShortBuffer?, count: Int): Boolean
  private external fun nativeFillLayoutSlab(
    handle: Long,
    slab: DoubleBuffer,
    capacityDoubles: Int,
    scrollOffset: Float,
    viewportHeight: Float,
    drawDistance: Float,
    outputScale: Float,
  ): Int
  private external fun nativeFillTypeStats(
    handle: Long,
    out: DoubleBuffer,
    capacityDoubles: Int,
    outputScale: Float,
  ): Int
  private external fun nativeSetItemSize(handle: Long, index: Int, size: Float): Boolean
  private external fun nativeSetItemSizesBatch(
    handle: Long,
    pairs: DoubleBuffer,
    pairCount: Int,
    scale: Float,
  ): Boolean
  private external fun nativeSetItemSizesBatchAnchored(
    handle: Long,
    pairs: DoubleBuffer,
    pairCount: Int,
    scale: Float,
    anchorIndex: Int,
  ): Float
  private external fun nativeSeedTypeMeans(
    handle: Long,
    pairs: DoubleBuffer,
    pairCount: Int,
    scale: Float,
  ): Boolean
  private external fun nativeResetItemSizes(handle: Long): Boolean
  private external fun nativeRemapItemSizes(
    handle: Long,
    pairs: DoubleBuffer,
    pairCount: Int,
  ): Boolean
  private external fun nativeResetAll(handle: Long)
  private external fun nativeGetMemoryFootprint(handle: Long): Long
  private external fun nativeGetTotalSize(handle: Long): Float
  private external fun nativeGetOffset(handle: Long, index: Int): Float
  private external fun nativeGetSize(handle: Long, index: Int): Float
  private external fun nativeGetLayoutVersion(handle: Long): Int
  private external fun nativeGetEngagedRange(
    handle: Long,
    scrollOffset: Float,
    viewportHeight: Float,
    drawDistance: Float,
    out: IntArray,
  )

  companion object {
    init {
      System.loadLibrary("NitroList")
    }
  }
}
